src/main/java/org/terasology/additionalRails/action/CargoCartAction.java

```diff
@@ -0,0 +1,51 @@
+package org.terasology.additionalRails.action;
+
+import org.terasology.additionalRails.components.CargoCartComponent;
+import org.terasology.entitySystem.entity.EntityManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterMode;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.entitySystem.systems.UpdateSubscriberSystem;
+import org.terasology.logic.inventory.InventoryComponent;
+import org.terasology.logic.inventory.ItemComponent;
+import org.terasology.math.geom.Vector3f;
+import org.terasology.minecarts.components.RailVehicleComponent;
+import org.terasology.registry.In;
+
+@RegisterSystem(RegisterMode.AUTHORITY)
+public class CargoCartAction extends BaseComponentSystem implements UpdateSubscriberSystem {
+    
+    private static final int MAX_ITEMS = 2971;
+    
+    @In
+    private EntityManager entityManager;
+    
+    @Override
+    public void update(float delta) {
```

Comment:
Do you need to run this method every frame or can you instead run it every time an item is added or removed?

Reply:
I'm also not sure how the Velocity stuff works, but it may be worth it to store a weight somewhere, rather than recalculating it every time.

Reply:
I'll try that.

Reply:
I changed it so that it stores the weight in the CargoCartComponent and uses that to recalculate the velocity each frame, it seems to work fine.